### DISCUSSION TITLE
Plotting the time index for a Pandas Dataframe?

### DISCUSSION THREAD

**Discussion by spchamp** [pyqtgraph/pyqtgraph#2891] [Q&A] [answered]
Has anyone had any experience with plotting using pyqtgraph, for a Pandas dataframe using a DatetimeIndex?

In the following example, I'll use [yfinance](https://pypi.org/project/yfinance/) as it's much easier to install and set up, juxtaposed to [something I've been hacking on](https://github.com/fxtatools/pyfx_dispatch_oanda/blob/main/src/pyfx/dispatch/oanda/models/get_instrument_candles200_response.py)

With the following - using pyqtgraph together with yfinanace - the plot will display three days of  OHLC candlesticks quotes for the FX instrument, USDJPY, at 1 minute time frame/granularity for the quotes. 

I'm hoping to figure out how to provide a matching time index for the chart, using the dataframe index or dataframe columns as an iteration source. Of course, these values could be readily converted to any number of timestamp strings, e.g iso format or display format, and can be easily converted to a series of ordinal index values.

I've taken a look at the [customPlot example](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/customPlot.py). For the second example, I'm not sure if the indexing works quite the same as when using a GraphicsObject directly?

```python
# qtfx_yf.py - prototypes for Qt integration, using yfinance

import numpy as np
import pyqtgraph as pg
import pandas as pd
import qtpy.QtGui as QtGui
import qtpy.QtCore as QtCore
from qtpy.QtWidgets import QApplication
from typing import TYPE_CHECKING


class QuotesView(pg.GraphicsObject):
    ## based on example pyqtgraph.examples.customGraphicsItem
    
    if TYPE_CHECKING:
        df: pd.DataFrame
    
    def __init__(self, df: pd.DataFrame):
        pg.GraphicsObject.__init__(self)
        # dataframe could be reused for indicator plots - average true range, etc.
        # self.df = df
        self.generatePicture(df)
        
    def generatePicture(self, df: pd.DataFrame):
        self.picture = QtGui.QPicture()
        p = QtGui.QPainter(self.picture)

        bulls_brush = pg.mkBrush('g')
        bears_brush = pg.mkBrush('r')
        bulls_pen = pg.mkPen('g')
        bears_pen = pg.mkPen('r')
        
        # df.reset_index(inplace=True)
        df = df.reset_index()
        
        w = 0.4

        def draw(row: pd.Series):
            idx = row.name
            close = row['Close']
            open = row['Open']
            if open > close:  # should be: open < close (screenshot shows the value as here)
                p.setBrush(bulls_brush)
                p.setPen(bulls_pen)  
            else:
                p.setBrush(bears_brush)
                p.setPen(bears_pen)  
            p.drawLine(QtCore.QPointF(idx, row['Low'],), QtCore.QPointF(idx, row['High']))
            p.drawRect(QtCore.QRectF(idx-w, open, w*2, close-open))
            return row
        df.apply(draw, axis=1)
        p.end()   
        
    ## from the pyqtgraph example
    
    def boundingRect(self):
        return QtCore.QRectF(self.picture.boundingRect())
    
    def paint(self, p, *args):
        p.drawPicture(0, 0, self.picture)
    

if __name__ =="__main__":

    import yfinance as yf

    app: QApplication = QApplication.instance() or QApplication([])

    df: pd.DataFrame = yf.download("USDJPY=X", interval="1m", period="3D")
    dt_index = df.index
    graph = QuotesView(df)

    ## [?]
    dtaxis = pg.DateAxisItem(orientation = 'bottom', utcOffset=0)
    wdgt = pg.PlotWidget(axisItems = dict(bottom = dtaxis))
    wdgt.setWindowTitle("%s - %s" % (df.index[0], df.index[-1]))

    # xvals = dt_index.to_numpy(copy=False, dtype=np.int64) // 10**9
    
    wdgt.addItem(graph)

    ## [?] tried a few approaches here
    # wdgt.plot(x=dt_index)
    # wdgt.plot(x=pd.to_datetime(dt_index)) 
    wdgt.plot() 

    wdgt.show()  
    app.exec()
```

The input dataframe looks something like the following:
```
In [13]: df
Out[13]: 
                                 Open        High         Low       Close   Adj Close  Volume
Datetime                                                                                     
2023-11-27 00:00:00+00:00  149.576004  149.578995  149.537994  149.544998  149.544998       0
2023-11-27 00:01:00+00:00  149.542999  149.542999  149.492004  149.509003  149.509003       0
2023-11-27 00:02:00+00:00  149.516998  149.526001  149.494003  149.507004  149.507004       0
2023-11-27 00:03:00+00:00  149.509003  149.550995  149.505997  149.550003  149.550003       0
2023-11-27 00:04:00+00:00  149.550003  149.557007  149.544006  149.550003  149.550003       0
...                               ...         ...         ...         ...         ...     ...
2023-11-29 19:29:00+00:00  147.233994  147.242996  147.216995  147.231995  147.231995       0
2023-11-29 19:30:00+00:00  147.227997  147.248001  147.218002  147.246002  147.246002       0
2023-11-29 19:31:00+00:00  147.244995  147.270004  147.240997  147.255997  147.255997       0
2023-11-29 19:32:00+00:00  147.257996  147.264999  147.242004  147.248001  147.248001       0
2023-11-29 19:33:00+00:00  147.244003  147.244003  147.244003  147.244003  147.244003       0

[4050 rows x 6 columns]

In [14]: df.info()
<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 4050 entries, 2023-11-27 00:00:00+00:00 to 2023-11-29 19:33:00+00:00
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Open       4050 non-null   float64
 1   High       4050 non-null   float64
 2   Low        4050 non-null   float64
 3   Close      4050 non-null   float64
 4   Adj Close  4050 non-null   float64
 5   Volume     4050 non-null   int64  
dtypes: float64(5), int64(1)
memory usage: 221.5 KB
```

With this small bit of source code, pyqtgraph displays all the quote prices correctly. It runs pretty quickly too. The Y axis is just right

![image](https://github.com/pyqtgraph/pyqtgraph/assets/1203789/c15cad5b-5ea7-47ee-b0f3-57130535131e)

After trying a few different approaches, I haven't been able to figure out how to get the timestamps from the dataframe to be displayed on the chart though. The X axis in the graph has stayed pretty much the same, from "Thu 01" to about an hour later.

Would there be any advice here?

For the time index in this chart, it may need a custom subclass of DateAxisItem?

**Comment by spchamp** [comment]
Updated example
- adding a pen, brush, and value test for `open == close`
- using the "bullish trend" pen when `open < close`

```ipython
# qtfx_yf.py - prototypes for Qt integration, using yfinance

import numpy as np
import pyqtgraph as pg # type: ignore[import-untyped]
import pandas as pd
import qtpy.QtGui as QtGui
import qtpy.QtCore as QtCore
from qtpy.QtWidgets import QApplication
from typing import TYPE_CHECKING


class QuotesView(pg.GraphicsObject):
    ## after  pyqtgraph.examples.customGraphicsItem
    
    if TYPE_CHECKING:
        df: pd.DataFrame
    
    def __init__(self, df: pd.DataFrame):
        pg.GraphicsObject.__init__(self)
        # self.df = df
        self.generatePicture(df)
        
    def generatePicture(self, df: pd.DataFrame):
        self.picture = QtGui.QPicture()
        p = QtGui.QPainter(self.picture)

        bulls_brush = pg.mkBrush('g')
        bears_brush = pg.mkBrush('r')
        bulls_pen = pg.mkPen('g')
        bears_pen = pg.mkPen('r')
        nochg_brush = pg.mkBrush('y')
        nochg_pen = pg.mkPen('y')
        
        # df.reset_index(inplace=True)
        df = df.reset_index()
        
        w = 0.4
        def draw(row: pd.Series):
            idx = row.name
            close = row['Close']
            open = row['Open']
            if open == close:
                p.setBrush(nochg_brush)
                p.setPen(nochg_pen)  
            elif open < close:
                p.setBrush(bulls_brush)
                p.setPen(bulls_pen)  
            else:
                p.setBrush(bears_brush)
                p.setPen(bears_pen)  
            p.drawLine(QtCore.QPointF(idx, row['Low'],), QtCore.QPointF(idx, row['High']))
            p.drawRect(QtCore.QRectF(idx-w, open, w*2, close-open))
            return row
        df.apply(draw, axis=1)
        p.end()   
        
    
    ## from the pyqtgraph example
    
    def boundingRect(self):
        return QtCore.QRectF(self.picture.boundingRect())
    
    
    def paint(self, p, *args):
        p.drawPicture(0, 0, self.picture)
    

if __name__ =="__main__":

    import yfinance as yf  # type: ignore[import-untyped]

    app: QApplication = QApplication.instance() or QApplication([])

    df: pd.DataFrame = yf.download("USDJPY=X", interval="1m", period="3D")
    dt_index = df.index
    graph = QuotesView(df)

    ## [?]
    dtaxis = pg.DateAxisItem(orientation = 'bottom', utcOffset=0)
    wdgt = pg.PlotWidget(axisItems = dict(bottom = dtaxis))
    wdgt.setWindowTitle("%s - %s" % (df.index[0], df.index[-1]))

    # xvals = dt_index.to_numpy(copy=False, dtype=np.int64) // 10**9
    
    wdgt.addItem(graph)
    # wdgt.plot(x=dt_index)
    # wdgt.plot(x=pd.to_datetime(dt_index))
    
    wdgt.plot()
    wdgt.show()  
    app.exec()
```

Example display, showing more detail when `open == close` at a certain time ..

![image](https://github.com/pyqtgraph/pyqtgraph/assets/1203789/1e382e47-ce89-49bc-a291-3f797d233e53)

**Comment by spchamp** [comment, marked as answer]
Adding this to the `generatePicture()` call, it seems to have worked out - converting the index values to epoch integer timestamps  during `generatePicture()`

```python
        df = df.copy()
        df.index = dt_index.to_numpy(copy=False, dtype=np.int64) // 10**9
```

It seems that it _can use_ an epoch-integer timestamp within the DataFrame.apply() call, after all.

Updated example
```python
# qtfx_yf.py - prototypes for Qt integration, using yfinance

import numpy as np
import pyqtgraph as pg  # type: ignore[import-untyped]
import pandas as pd
import qtpy.QtGui as QtGui
import qtpy.QtCore as QtCore
from qtpy.QtWidgets import QApplication
from typing import TYPE_CHECKING


class QuotesView(pg.GraphicsObject):
    ## based on pyqtgraph.examples.customGraphicsItem

    if TYPE_CHECKING:
        df: pd.DataFrame

    def __init__(self, df: pd.DataFrame):
        pg.GraphicsObject.__init__(self)
        # self.df = df
        self.generatePicture(df)

    def generatePicture(self, df: pd.DataFrame):
        self.picture = QtGui.QPicture()
        p = QtGui.QPainter(self.picture)

        df = df.copy()
        ## source for the optimized pd.Timestamp to epoch timestamp conversion:
        ## https://stackoverflow.com/a/54313505/1061095
        df.index = dt_index.to_numpy(copy=False, dtype=np.int64) // 10**9
    
        w = (df.index[1] - df.index[0]) / 3
        pw = np.trunc(w / 9) # pen width ...

        bulls_brush = pg.mkBrush('g')
        bulls_pen = pg.mkPen('g', width=pw)
        bears_brush = pg.mkBrush('r')
        bears_pen = pg.mkPen('r', width=pw)
        nochg_brush = pg.mkBrush('y')
        nochg_pen = pg.mkPen('y', width=pw)

        def draw(row: pd.Series):
            idx = row.name
            close = row['Close']
            open = row['Open']
            if open == close:
                p.setBrush(nochg_brush)
                p.setPen(nochg_pen)
            elif open < close:
                p.setBrush(bulls_brush)
                p.setPen(bulls_pen)
            else:
                p.setBrush(bears_brush)
                p.setPen(bears_pen)
            p.drawLine(QtCore.QPointF(idx, row['Low'],), QtCore.QPointF(idx, row['High']))
            p.drawRect(QtCore.QRectF(idx-w, open, w*2, close-open))
            return row
        df.apply(draw, axis=1)
        p.end()

    ## from the pyqtgraph example

    def boundingRect(self):
        return QtCore.QRectF(self.picture.boundingRect())

    def paint(self, p, *args):
        p.drawPicture(0, 0, self.picture)

        

if __name__ == "__main__":

    import yfinance as yf  # type: ignore[import-untyped]

    app: QApplication = QApplication.instance() or QApplication([])

    df: pd.DataFrame = yf.download("USDJPY=X", interval="1m", period="3D")
    dt_index = df.index
    graph = QuotesView(df)

    dtaxis = pg.DateAxisItem(orientation='bottom', utcOffset=0)     
    wdgt: pg.PlotWidget = pg.PlotWidget(axisItems=dict(bottom=dtaxis))
    wdgt.setWindowTitle("%s - %s" % (df.index[0], df.index[-1]))
    wdgt.addItem(graph)

    wdgt.plot()
    wdgt.show()
    app.exec()
```

Display :
![image](https://github.com/pyqtgraph/pyqtgraph/assets/1203789/1c5052f1-9973-4a37-b29e-2c64e3cbfcc5)

The huge spike in the market rate, there, is pretty tricky to find visually. It would not be as tough to find, if there was an open trade at that time lol. oof. Peculiar thing though, the market rate was the same at the 'open' and 'close' rates in that minute.

With a thin line style, it almost disappears from the graph - more to the usefulness of the pyqtgraph examples, such as for indicators that can be added to the graph axis.

**Comment by j9ac9k** [comment]
hi @spchamp 

Sorry you never got an answer from us, the `DateAxisItem` was a feature added by a contributor, and I haven't used it at all myself, so I wasn't in a position to provide much input.

That said, if there are changes to `DateAxisItem` that can be made to make it easier to integrate with pandas date time index, I would welcome that change into the library.  One of the long term goals I've had is to have better support for integration with pandas objects.

**Reply by spchamp** [reply]
No worries! I think it's an interesting feature. 

I think I was struggling a little, initially, with figuring out how to present the timestamps in the input data. Starting with the FX quotes data from yfinance, of course the initial dataframe uses a Pandas datetime index. This index uses pandas.Timestamp values. I think it's something like a series of Python-compatible datetime objects. Of course it's a different data type than numpy datetime64.

Converting the timestamp index to something like an epoch timestamp series - truncated to the integer part - for purpose of graph display, somehow  this just works? It seems that the DataAxisItem  interpolates the timestamp strings and display alignment, without further coordination needed for the input data series. I think it's really interesting how simple it is, at the point of application, given a suitably formatted timestamp series lol.

I think there may be a next challenge in figuring out how to keep the data updated in the graph, for any new data in the time series, when trying to use a dataframe as the iteration source. I noticed the notes in the [scrollingPlots example](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/examples/scrollingPlots.py), thinking to try a chunks-oriented approach, coordinated with some kind of a data fetch procedure for "new quotes", then something to merge the new data with any existing dataframe before updating the graph.

This adaptation onto the customGraphicsItem example, with its aggregate data series in the "Candlesticks" style, it doesn't use any PlotDataset or PlotDataItem objects though? I wonder if it may be missing features that may be available in the API, for literal plot curve objects. Now with maybe a little more of an understanding of the data layer, looking forward to some more tinkering though lol.

I'm really impressed with the extent of the examples, sincerely. imo it seems wildly more straightforward than [CLIM](https://www.lispworks.com/documentation/lww42/CLIM-W/html/climguide.htm), not to throw shade though lol, CLIM has its Common Lisp object model

**Reply by pijyoi** [reply]
I believe `pg.DateAxisItem` supports fractional seconds.
So you could change the lines
`df.index = dt_index.to_numpy(copy=False, dtype=np.int64) // 10**9`
`w = (df.index[1] - df.index[0]) / 3`
`idx = row.name`
to
`df.index = dt_index.to_numpy(copy=False, dtype=np.int64)`   # leave as nanoseconds
`w = (df.index[1] - df.index[0]) / 1e9 / 3`  # set width to 1/3 of a second
`idx = row.name / 1e9`  # convert to seconds w/o truncation. (and maybe use a different variable name than `idx` as `idx` implies integer)

**Comment by Denfrommur** [comment]
Good afternoon. Can you tell me what I'm doing wrong? The x-axis always shows timestamp without converting it to a date.

Form, Window = uic.loadUiType("Isee_1.ui")

app = QApplication([])
window = Window()

def compare(iday_2):

    iday_2 = iday_difference.tail(200).Close
    #df_reset = iday_2.reset_index()


    xx =  df_reset.Minute.astype(np.int64)//10**9
    #xx = pd.to_datetime(df_reset.Minute)
    date_axis = TimeAxisItem(orientation='bottom')

    iday_2.info()
    form.graphicsView.clear()

    form.graphicsView.plot(x=[x.timestamp() for x in iday_2.index], y=iday_2.Close, axisItem={'bottom': date_axis})

form.pushButton.clicked.connect(compare)

app.exec()

1771272960
<class 'pandas.core.series.Series'>
DatetimeIndex: 200 entries, 2026-02-16 20:16:00 to 2026-02-16 23:49:00
Series name: Close
Non-Null Count  Dtype  
--------------  -----  
200 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB
<img width="1371" height="831" alt="pict" src="https://github.com/user-attachments/assets/43424652-633d-4f32-ad49-a37080619ec9" />
